### DISCUSSION TITLE
HELP! CountVectorizer and ClassTfidfTransformer changing clustering results

### DISCUSSION THREAD

**Discussion by jdweaver14** [MaartenGr/BERTopic#1192] [Q&A]
Hi everyone, 

I apologize for not posting images here, but I am not at liberty to share the data I am working with. 

I am new to BERTopic, but I was under the impression that the ClassTfidfTransformer and CountVectorizer steps were only used after embeddings (that did not use preprocessing steps) and clustering to improve topic representation. However, when I run the same dataset with fixed random_state through the model with and without either or both of these steps introduced, I get both a different number of clusters and difference counts per cluster. This should not be happening if these steps are only used on the already calculated cluster information. Please help me understand what I am missing!

**Comment by GeorgeDittmar** [comment]
Yeah I see something similar with my own data and was about to ask if CountVectorizer etc can affect how clusters are generated when I am training the model. my understanding was similar to yours but maybe I misread the docs. If I run them as an update step the clusters stay the same but explicitly calling them in the BERTopic object setup seems to produce differing results.

**Reply by jdweaver14** [reply]
Agreed, I also found precomputed embeddings keeps clusters consistent, but not if an embedding model is defined in addition to the embeddings

**Comment by MaartenGr** [comment]
@jdweaver14 Could you share your full code? That allows me to see what exactly is happening here. Please make sure to share everything up to and including fitting the model. 

> I am new to BERTopic, but I was under the impression that the ClassTfidfTransformer and CountVectorizer steps were only used after embeddings (that did not use preprocessing steps) and clustering to improve topic representation. 

The CountVectorizer and ClassTfidfTransformer are indeed steps that are used **after** embedding the documents, dimensionality reduction, and clustering. So they should not influence the clustering themselves. Most likely, the change in clusters is happening somewhere else in the code.

**Reply by GeorgeDittmar** [reply]
So this is the code I use to build my topic models and if I pass in the vectorizer_model or representation_model when instantiating BERTopic, I get different cluster results than if I don't pass them in and I just let bertopic handle it however by default. More specifically I notice that one topic suddenly becomes very large if I pass those in. So I guess the question is are these two things really only used after the clusters are built or do they play a role somehow in the final cluster output beyond just creating the topic representations?
```
nr_topics = 'auto'
calc_probabilities = True
n_gram_range = (1,3)
min_topic_size = 20
top_n_words = 10

representation_model = KeyBERTInspired()
umap_model = UMAP(random_state=2022)
vectorizer_model = CountVectorizer(stop_words="english")

topic_model = BERTopic(top_n_words=top_n_words, embedding_model=sentence_model, umap_model=umap_model, nr_topics=nr_topics, calculate_probabilities=calc_probabilities, n_gram_range=n_gram_range, min_topic_size=min_topic_size)

topics, probs = topic_model.fit_transform(raw_data.utterance, embeddings=embeddings)```

**Reply by GeorgeDittmar** [reply]
I also am seeing a different number of clusters depending on if I am using the vectorizor and or representation model so it seems like it does play into how clusters are generated?

**Reply by MaartenGr** [reply]
Ah, now I see. The CountVectorizer and ClassTfidfTransformer do not influence the initial cluster creation as they are only created **after** the initial clusters have been formed through the cluster model. However, when you use `nr_topics="auto"` the low-level clusters that are initially created are further reduced based on the topic representation of those clusters. As a result, if you change the cluster representation with the vectorizer/transformer, then that will influence how the low-level clusters are merged. 

In other words, if you do not set `nr_topics` then it will not influence the clusters that are generated.

**Reply by GeorgeDittmar** [reply]
Ok that makes sense then. Is it possible to just run the representation stuff as a secondary step after it returns me the clusters or would I have to do that manually as something outside of bertopic? For example if I want to keep nr_topics=auto but then run the representation over that?

**Reply by MaartenGr** [reply]
Yes, that is what [`.update_topics`](https://maartengr.github.io/BERTopic/getting_started/topicrepresentation/topicrepresentation.html#update-topic-representation-after-training) is for. There, you can define different representation models to fine-tune the topic representations.

**Reply by GeorgeDittmar** [reply]
awesome thank you! I will test that out on my data now!
